File: core/src/main/java/moe/dituon/petpet/core/transition/RotateTransition.java
package moe.dituon.petpet.core.transition;

import lombok.Builder;
import lombok.Getter;

@Builder
@Getter
public class RotateTransition {
    public static final RotateTransition LINER = createDefaultInstance(EasingKeyword.LINEAR.function);
    public static final RotateTransition EASE = createDefaultInstance(EasingKeyword.EASE.function);
    public static final RotateTransition EASE_IN = createDefaultInstance(EasingKeyword.EASE_IN.function);
    public static final RotateTransition EASE_OUT = createDefaultInstance(EasingKeyword.EASE_OUT.function);
    public static final RotateTransition EASE_IN_OUT = createDefaultInstance(EasingKeyword.EASE_IN_OUT.function);
    public static final RotateTransition DEFAULT = LINER;

    public final float start;
    public final float end;
    public final EasingFunction easing;
    public final int rotateCount;

    public RotateTransition(float start, float end, EasingFunction easing, int rotateCount) {
        this.start = start;
        this.end = end + 360 * rotateCount;
        this.easing = easing;
        this.rotateCount = (int) ((end - start) / 360f);
    }

    public boolean isEmpty() {
        return false;
    }

    public float getAngle(int length, int index) {
        return this.easing.easingFrames(start, end, length + 1, index);
    }

    public static RotateTransition createDefaultInstance(EasingFunction function) {
        return RotateTransition.builder()
                .start(0).end(0)
                .rotateCount(1)
                .easing(function)
                .build();
    }

    public static RotateTransition empty() {
        return EmptyRotateTransition.INSTANCE;
    }

    @Override
    public String toString() {
        return "RotateTransition{" +
                "start=" + start +
                ", end=" + end +
                ", easing=" + easing +
                '}';
    }

    protected static class EmptyRotateTransition extends RotateTransition {
        public static final EmptyRotateTransition INSTANCE = new EmptyRotateTransition();

        public EmptyRotateTransition() {
            super(0, 0, LinearEasingFunction.INSTANCE, 0);
        }

        @Override
        public boolean isEmpty() {
            return true;
        }

        @Override
        public float getAngle(int length, int index) {
            return 0;
        }
    }
}

File: core/src/main/kotlin/moe/dituon/petpet/template/fields/transition/RotateTransition.kt
package moe.dituon.petpet.template.fields.transition

import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.*
import moe.dituon.petpet.core.transition.EasingFunction
import moe.dituon.petpet.core.transition.EasingKeyword
import moe.dituon.petpet.core.transition.RotateTransition

@Serializable
data class RotateTransitionSurrogate(
    val easing: EasingElement = EasingKeyword.LINEAR.function,
    val start: Float = 0f,
    val end: Float = 360f,
    @SerialName("rotate_count")
    val rotateCount: Int = 1,
)

object RotateTransitionSerializer : KSerializer<RotateTransition> {
    override val descriptor: SerialDescriptor = RotateTransitionSurrogate.serializer().descriptor

    override fun serialize(encoder: Encoder, value: RotateTransition) {
        val surrogate = RotateTransitionSurrogate(
            start = value.start,
            end = value.end,
            rotateCount = value.rotateCount,
            easing = value.easing
        )
        encoder.encodeSerializableValue(RotateTransitionSurrogate.serializer(), surrogate)
    }

    override fun deserialize(decoder: Decoder): RotateTransition =
        when (val element = decoder.decodeSerializableValue(JsonElement.serializer())) {
            is JsonPrimitive -> if (element.isString) when (element.content) {
                EasingKeyword.LINEAR.name -> RotateTransition.LINER
                EasingKeyword.EASE_IN.name -> RotateTransition.EASE_IN
                EasingKeyword.EASE_OUT.name -> RotateTransition.EASE_OUT
                EasingKeyword.EASE_IN_OUT.name -> RotateTransition.EASE_IN_OUT
                else -> RotateTransition.createDefaultInstance(EasingFunction.fromString(element.content))
            } else when (element.booleanOrNull) {
                true -> RotateTransition.DEFAULT
                false -> RotateTransition.empty()
                null -> throw IllegalArgumentException("Invalid rotate transition value: ${element.content}")
            }

            is JsonArray -> throw IllegalArgumentException("Rotate transition object can't decode via array")
            is JsonObject -> {
                val s = decoder.decodeSerializableValue(RotateTransitionSurrogate.serializer())
                RotateTransition.builder()
                    .rotateCount(s.rotateCount)
                    .start(s.start)
                    .end(s.end)
                    .easing(s.easing)
                    .build()
            }
        }
}

typealias RotateTransitionElement = @Serializable(RotateTransitionSerializer::class) RotateTransition